### PR TITLE
feat: Add `saveEventually` and `deleteEventually` in `ParseObject`

### PR DESCRIPTION
**Pull request by mbfakourii**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

## Issue

Closes: #149 and #194

## Approach

In this feature, Functions `saveEventually` and `deleteEventually` in `ParseObject` were added

---> in `saveEventually` 
At first, normal ```save``` command is entered in parseObject, if a ```NetworkError``` is encountered, this object is saved in a list with a key with the help of ParseCoreData.
In our collection, we have a list of objects in ```ParseCoreData``` that are in ```NetworkError```.

---> in `deleteEventually` 
At first, normal ```delete``` command is entered in parseObject, if a ```NetworkError``` is encountered, this object is saved in a list with a key with the help of ParseCoreData.
In our collection, we have a list of objects in ```ParseCoreData``` that are in ```NetworkError```.

---> in ```await ParseObject.submitEventually()```
With the help of the function ```await ParseObject.submitEventually()```  (which is debatable how to use it) we checks the error objcets

In this function, we first get the error parseObjects from ```ParseCoreData```, then we send this list to ```_saveChildren```. In this function, these objects are sent to the server in the form of chunks, and finally, we empty the ```ParseCoreData``` list.

In the following, we also take the list of ```parseObjects``` that must be deleted from ```ParseCoreData``` and try to delete them, and then we delete the successfully deleted parseObjects from the ```ParseCoreData``` list.

@parse-community/flutter-sdk-review 
Please check these two functions thoroughly and give suggestions, this PR needs discussion.
and please check the algorithm, is this method good?

## Tasks

- [x] Add tests
- [ ] Add changes to documentation (guides, repository pages, code comments)

### CODE REVIEW

**Comment by mtrezza** [comment]
> Please check these two functions thoroughly and give suggestions, this PR needs discussion.

Changed to draft

**Comment by mbfakourii** [comment]
@mtrezza @Nidal-Bakir 
Do you think the implemented method is appropriate?

**Comment by Nidal-Bakir** [comment]
I suggest looking at what other SDKs do and doing the same. There is the concept of a Queue so the objects get saved in order. 

Do other SDKs save the unsaved Objects even when the app is closed(destroyed) using native services (android and iOS) when the connection is restored for example?

**Comment by mbfakourii** [comment]
> I suggest looking at what other SDKs do and doing the same. There is the concept of a Queue so the objects get saved in order.
> 
> Do other SDKs save the unsaved Objects even when the app is closed(destroyed) using native services (android and iOS) when the connection is restored for example?

I think the queue issue happens in [dio](https://pub.dev/documentation/dio/latest/dio/QueuedInterceptor-class.html) or Http and we don't need to create this queue.

According to the ios documentation

[saveEventually](https://parseplatform.org/Parse-SDK-iOS-OSX/api/Classes/PFObject.html#/c:objc(cs)PFObject(im)saveEventually)
```
Saves this object to the server at some unspecified time in the future, even if Parse is currently inaccessible.

Use this when you may not have a solid network connection, and don’t need to know when the save completes. If there is some problem with the object such that it can’t be saved, it will be silently discarded.

Objects saved with this method will be stored locally in an on-disk cache until they can be delivered to Parse. They will be sent immediately if possible. Otherwise, they will be sent the next time a network connection is available. Objects saved this way will persist even after the app is closed, in which case they will be sent the next time the app is opened. If more than 10MB of data is waiting to be sent, subsequent calls to -saveEventually will cause old saves to be silently discarded until the connection can be re-established, and the queued objects can be saved.
```

Based on what I understand, the internal structure of Android is implemented with this concept. If possible, take a look at the [Android structure](https://github.com/parse-community/Parse-SDK-Android/blob/2bf43513973277558a00e55ad96d0f5b33baeb5b/parse/src/main/java/com/parse/ParseObject.java#L2339).

The concept that I didn't understand was where to double check and send the saved it objects?

I wrote the codes of this section in the ```await ParseObject.submitEventually()``` function! And I allow the user to use it wherever they need. But not sure about that !

**Comment by mbfakourii** [comment]
@parse-community/android-sdk and @parse-community/ios-sdk

It is possible for you to participate in this discussion

File: packages/dart/lib/src/objects/parse_object.dart [hunk 1]
```diff
@@ -677,4 +722,89 @@ class ParseObject extends ParseBase implements ParseCloneable {
       return this;
     }
   }
+
+  static Future<ParseResponse?> submitEventually(
```
Comment (Nidal-Bakir) [inline packages/dart/lib/src/objects/parse_object.dart:726]: who should call this faction the end-user or the SDK itself when initialized?

**Review by Nidal-Bakir — COMMENTED** [review]
Overall everything is good.

I have some thoughts:
- What will happen if the current user logs out?
  -  should we clear the cache? or could you give the end user the option to decide what should happen?
- should we give the end user control over the cached objects that need to be saved/deleted?

File: packages/dart/lib/src/objects/parse_object.dart [hunk 2]
```diff
@@ -677,4 +722,89 @@ class ParseObject extends ParseBase implements ParseCloneable {
...
+    List<String>? listSaves =
+        await coreStore.getStringList(keyParseStoreObjects);
+
+    if (listSaves != null) {
+      List<ParseObject> parseObjectList = [];
+      for (var element in listSaves) {
+        // decode json
+        dynamic object = json.decode(element);
+        parseObjectList
+            .add(ParseObject(object[keyVarClassName]).fromJson(object));
```
[the remark this answers is not in the document]
Comment (mbfakourii) [reply packages/dart/lib/src/objects/parse_object.dart:749]: ClassName is not used in the fromJson function, I had to write it like this

**Comment by mbfakourii** [comment]
> Overall everything is good.
> 
> I have some thoughts:
> 
> * What will happen if the current user logs out?
>   
>   * should we clear the cache? or could you give the end user the option to decide what should happen?
> * should we give the end user control over the cached objects that need to be saved/deleted?

At the moment, the caches are not deleted, but I think a tool should be written to delete the cache

Regarding the second question, I did not see anything in other SDKs. Do you have any suggestions for this? What control do you mean?

**Comment by Nidal-Bakir** [comment]
> At the moment, the caches are not deleted, but I think a function should be written to delete the cache

Got it.


> Regarding the second question, I did not see anything in other SDKs. Do you have any suggestions for this? What control do you mean?

I mean what if the user wants to clear the currently cached objects for saving/deleting and start over?
Should we give expose such a feature?

**Comment by mbfakourii** [comment]
> I mean what if the user wants to clear the currently cached objects for saving/deleting and start over? Should we expose such a feature?
> 
> Or in general, should we expose the current cache state so the end user can alter it?
> 
> If yes we should consider a saving mechanism like functions to be invoked. for example:
> 
> * onBeforStartSaving()
> * shouldSave()
> * shouldDelete()
> * onBeforSave()
> * onBeforDelete()
> * onAfterSave()
> * onAfterDelete()
> * onAllSaved()
> * onAllDeleted()
> 
> something like that

I don't think this is necessary. I haven't seen anything about this in other SDKs. Do you have a reason why it needs to be added?

**Comment by Nidal-Bakir** [comment]
No, I'm just wounding if we need to give the end user finer control over the eventual save in general. If the other SDKs do not do that we are good.

So currently, we need to discuss how the submitEventually function should be called and what will happen when the user logs out when there are some unsaved objects.

**Comment by mbfakourii** [comment]
> > Do you have a reason why it needs to be added?
> 
> No, I'm just wounding if we need to give the end user finer control over the eventual save in general. If the other SDKs do not do that we are good.
> 
> So currently, we need to discuss how the `submitEventually` function should be called and what will happen when the user logs out when there are some unsaved objects.

Do you think we should leave this function to the user or place it somewhere like the `save()` function?

**Comment by mtrezza** [comment]
What's missing in this PR to get it merged?

**Comment by mbfakourii** [comment]
> What's missing in this PR to get it merged?

The code is finished, the tests must be write

**Comment by mbfakourii** [comment]
The error related to [`Test Flutter beta`](https://github.com/parse-community/Parse-SDK-Flutter/actions/runs/6397441480/job/17365365740?pr=911) has been resolved in [this](https://github.com/parse-community/Parse-SDK-Flutter/pull/969) PR

File: packages/dart/lib/src/objects/parse_object.dart [hunk 3]
```diff
@@ -162,7 +162,7 @@ class ParseObject extends ParseBase implements ParseCloneable {
       return response;
     } else {
       // if have network connection error
-      if ((response.error?.message ?? "") == "NetworkError") {
+      if ((response.error?.message ?? "").contains("NetworkError")) {
```
Comment (mtrezza) [inline packages/dart/lib/src/objects/parse_object.dart:165]: Why is this parsing an error message, is there no error code to compare?

Comment (mbfakourii) [reply packages/dart/lib/src/objects/parse_object.dart]: I changed it, please check again

**Comment by mbfakourii** [comment]
@mtrezza 

Done 😃

File: packages/dart/lib/src/objects/parse_object.dart [hunk 4]
```diff
@@ -271,7 +290,7 @@ class ParseObject extends ParseBase implements ParseCloneable {
         final ParseResponse response = await batchRequest(
           requests,
           chunk,
-          client: _client,
+          client: client,
```
Comment (mtrezza) [inline packages/dart/lib/src/objects/parse_object.dart:293]:
```suggestion
          client,
```

File: packages/dart/lib/parse_server_sdk.dart [hunk 5]
```diff
@@ -27,62 +27,116 @@ export 'src/network/parse_dio_client.dart';
 export 'src/network/parse_http_client.dart';
 
 part 'src/base/parse_constants.dart';
+
```
Comment (mtrezza) [inline packages/dart/lib/parse_server_sdk.dart:30]: Are these empty lines an automatic lint change or were they added by accident?

**Comment by mtrezza** [comment]
Note for review: The Flutter test is failing, hence we need to merge https://github.com/parse-community/Parse-SDK-Flutter/pull/969 first before merging this PR.

File: packages/dart/lib/src/objects/parse_object.dart (hunk 4 above)
Comment (mbfakourii) [reply packages/dart/lib/src/objects/parse_object.dart:293]: This suggestion is not correct, it creates a conflict.

File: packages/dart/lib/parse_server_sdk.dart (hunk 5 above)
Comment (mbfakourii) [reply packages/dart/lib/parse_server_sdk.dart:30]: This has been generated using [auto code formatting](https://docs.flutter.dev/tools/formatting#automatically-formatting-code-in-android-studio-and-intellij) in Dart.

**Review by mtrezza — COMMENTED** [review]
Could you check whether I've resolved the conflicts correctly? Then I believe this can be merged.

**Comment by mbfakourii** [comment]
> Could you check whether I've resolved the conflicts correctly? Then I believe this can be merged.

I checked, there was no problem

File: packages/dart/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,9 @@
+## [6.2.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-6.1.0...dart-6.2.0) (2023-10-17)
```
Comment (mtrezza) [inline packages/dart/CHANGELOG.md:1]:
```suggestion
## [6.2.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-6.1.0...dart-6.2.0) (2023-10-18)
```

**Review by mtrezza — APPROVED** [review]

**Pull request merged**